src/main/java/uk/gov/hmcts/reform/unspec/callback/CallbackHandlerFactory.java

```diff
@@ -2,27 +2,39 @@
 
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.ccd.client.model.AboutToStartOrSubmitCallbackResponse;
 import uk.gov.hmcts.reform.ccd.client.model.CallbackResponse;
+import uk.gov.hmcts.reform.unspec.helpers.CaseDetailsConverter;
+import uk.gov.hmcts.reform.unspec.model.CaseData;
 
+import java.util.Arrays;
 import java.util.HashMap;
-import java.util.List;
 
 import static java.util.Optional.ofNullable;
 
 @Service
 public class CallbackHandlerFactory {
 
     private final HashMap<String, CallbackHandler> eventHandlers = new HashMap<>();
+    private final CaseDetailsConverter caseDetailsConverter;
 
-    @Autowired(required = false)
-    public CallbackHandlerFactory(List<CallbackHandler> beans) {
-        beans.forEach(bean -> bean.register(eventHandlers));
+    @Autowired
+    public CallbackHandlerFactory(CaseDetailsConverter caseDetailsConverter, CallbackHandler... beans) {
+        this.caseDetailsConverter = caseDetailsConverter;
+        Arrays.asList(beans).forEach(bean -> bean.register(eventHandlers));
     }
 
     public CallbackResponse dispatch(CallbackParams callbackParams) {
         String eventId = callbackParams.getRequest().getEventId();
         return ofNullable(eventHandlers.get(eventId))
-            .map(h -> h.handle(callbackParams))
+            .map(h -> processEvent(callbackParams, h))
             .orElseThrow(() -> new CallbackException("Could not handle callback for event " + eventId));
     }
+
+    private CallbackResponse processEvent(CallbackParams callbackParams, CallbackHandler handler) {
+        CaseData caseData = caseDetailsConverter.toCaseData(callbackParams.getRequest().getCaseDetails());
+        return handler.isEventAlreadyProcessed(caseData)
+            ? AboutToStartOrSubmitCallbackResponse.builder().build()
```

Comment:
As this will be invoked from UI callbacks as well, should we let the relevant handler return this? For UI callbacks we will want to populate error back.

Reply:
UI callbacks have camunda task id defaulted as DEFAULT so they will always go for Event Handler

Reply:
confused by this logic, why return an empty callback response if the event is already processed? Should it not return some error?